### PR TITLE
[WIP] A solution to prevent zombie containers locally and in CI

### PR DESCRIPTION
**Pull request by larroy**
## Description ##
This PR adds mechanisms on the build scripts to cleanup docker containers when there is a cancellation, either from the user by sending SIGINT / SIGTERM or when the process inside the container ends with an error.

It also propagates the environment variables from Jenkins which are used by the process tree killer to identify runaway processes so processes inside the container are killed when the Jenkins job is stopped.

With this patch we catch SIGINT and SIGTERM and install a handler on atexit which cleans (stops) docker containers which were created during execution of the build.

We also switch to python docker API for better management of running containers.

Since Jenkins is suppossed to send SIGINT / SIGTERM on job cancellation this should also provent zombie containers in CI.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: ci/build.py [hunk 1]
```diff
@@ -23,7 +23,11 @@
 """
 
 __author__ = 'Marco de Abreu, Kellen Sunderland, Anton Chernov, Pedro Larroy'
-__version__ = '0.1'
+__version__ = '0.8'
+
+def pip_install(pkg):
```
Comment (lebeg) [inline ci/build.py]: We shouldn't install anything during a build.py run, all requirements should be there already

File: ci/build.py [hunk 2]
```diff
@@ -37,11 +41,54 @@
...
 from typing import *
 from util import *
+try:
+    pip_install('docker')
+    import docker
+    import docker.models
+    import docker.errors
+    import signal
+    import atexit
+except ImportError as e:
```
Comment (lebeg) [inline ci/build.py]: A usual import error I think should be enough

File: ci/build.py [hunk 3]
```diff
@@ -37,11 +41,54 @@
...
+    def add_container(self, container: docker.models.containers.Container):
+        assert isinstance(container, docker.models.containers.Container)
+        self.containers.add(container)
+
+    def remove_container(self, container: docker.models.containers.Container):
+        assert isinstance(container, docker.models.containers.Container)
+        self.containers.remove(container)
+
+    def _cleanup_containers(self):
+        if self.containers:
```
Comment (lebeg) [inline ci/build.py]: Shouldn't we just return if not?

File: ci/build.py [hunk 4]
```diff
@@ -37,11 +41,54 @@
...
+            logging.warning("Cleaning up containers")
+        docker_client = docker.from_env()
+        try:
+            stop_timeout = int(os.environ.get("DOCKER_STOP_TIMEOUT", 3))
```
Comment (lebeg) [inline ci/build.py]: Could we have "3" in a named variable?

File: ci/build.py [hunk 5]
```diff
@@ -37,11 +41,54 @@
...
+        except Exception as e:
+            stop_timeout = 3
+        for container in self.containers:
+            try:
+                container.stop(timeout=stop_timeout)
+                logging.warning("☠: %s", trim_container_id(container.id))
```
Comment (lebeg) [inline ci/build.py]: Should we have a more meaningful warning text than ☠?

File: ci/build.py [hunk 6]
```diff
@@ -37,11 +41,54 @@
...
+            except Exception as e:
+                logging.exception(e)
```
Comment (lebeg) [inline ci/build.py]: That might be a good place for a warning

File: ci/build.py (hunk 5 above)
Comment (lebeg) [reply ci/build.py]: Do you think we need a warning level here?

File: ci/build.py [hunk 7]
```diff
@@ -121,7 +185,7 @@ def build_docker(platform: str, docker_binary: str, registry: str, num_retries:
     # container match permissions of the local user. Same for the group.
     #
     # These variables are used in the docker files to create user and group with these ids.
-    # see: docker/install/ubuntu_adduser.sh
+    # see: dockerfiles/install/ubuntu_adduser.sh
```
Comment (lebeg) [inline ci/build.py]: Let's not move the files in this change for now. We probably would like this change to be merged to release branches as well and we'd need to keep it as crisp as possible.

File: ci/build.py [hunk 8]
```diff
@@ -180,63 +245,133 @@ def default_ccache_dir() -> str:
...
+                  local_ccache_dir: str,
+                  cleanup: Cleanup,
                   dry_run: bool = False,
-                  interactive: bool = False) -> str:
+                  interactive: bool = False) -> int:
+    #
+    # Environment setup
+    #
+    environment = {
+        'CCACHE_MAXSIZE': '20G',
```
Comment (lebeg) [inline ci/build.py]: You've changed the default value of this which was 500G previously. This would delete almost the whole cache on the ci. It makes sense to check whether the script is run on the ci and if not, set a lower value. But I would leave it for a later change and just fix the value by using the variable `CCACHE_MAXSIZE`  which was previously there.

File: ci/build.py [hunk 9]
```diff
@@ -180,63 +245,133 @@ def default_ccache_dir() -> str:
...
+    # Equivalent command
+    docker_cmd_list = [docker_binary, 'run',
+               '--rm',
                '--shm-size={}'.format(shared_memory_size),
                '-v', "{}:/work/mxnet".format(mx_root),  # mount mxnet root
                '-v', "{}:/work/build".format(local_build_folder),  # mount mxnet/build for storing build artifacts
                '-v', "{}:/work/ccache".format(local_ccache_dir),
                '-u', '{}:{}'.format(os.getuid(), os.getgid()),
-               '-e', 'CCACHE_MAXSIZE={}'.format(CCACHE_MAXSIZE),
+               '-e', 'CCACHE_MAXSIZE=15G',
```
Comment (lebeg) [inline ci/build.py]: See comment above

File: ci/build.py [hunk 10]
```diff
@@ -180,63 +245,133 @@ def default_ccache_dir() -> str:
...
+        #
+        signal.pthread_sigmask(signal.SIG_BLOCK, {signal.SIGINT, signal.SIGTERM})
+        container = docker_client.containers.run(
+            tag,
+            detach=True,
+            command=command,
+            #auto_remove=True,
+            shm_size=shared_memory_size,
+            user='{}:{}'.format(os.getuid(), os.getgid()),
+            volumes={
```
Comment (lebeg) [inline ci/build.py]: This looks so much better!

File: ci/build.py [hunk 11]
```diff
@@ -180,63 +245,133 @@ def default_ccache_dir() -> str:
...
+        # Race condition:
+        # If the previous call is interrupted then it's possible that the container is not cleaned up
+        # We avoid by masking the signals temporarily
+        cleanup.add_container(container)
+        signal.pthread_sigmask(signal.SIG_UNBLOCK, {signal.SIGINT, signal.SIGTERM})
+        #
+        #############################
+
+        stream = container.logs(stream=True)
+        for chunk in stream:
```
Comment (lebeg) [inline ci/build.py]: Why is this necessary?

File: ci/build.py [hunk 12]
```diff
@@ -180,63 +245,133 @@ def default_ccache_dir() -> str:
...
+            sys.stdout.buffer.write(chunk)
+            sys.stdout.buffer.flush()
+        stream.close()
+        try:
+            logging.info("Waiting for status of container %s", trim_container_id(container.id))
+            wait_result = container.wait(timeout=10)
```
Comment (lebeg) [inline ci/build.py]: We should have a variable for the timeout value. In what units is it?

File: ci/build.py [hunk 13]
```diff
@@ -248,7 +383,11 @@ def load_docker_cache(tag, docker_registry) -> None:
     else:
         logging.info('Distributed docker cache disabled')
 
+
 def main() -> int:
+    import pprint as pp
```
Comment (lebeg) [inline ci/build.py]: Should we have it where all other imports are?

File: ci/dockerfiles/runtime_functions.sh [hunk 14]
```diff
@@ -305,6 +305,14 @@ build_ubuntu_cpu() {
     build_ubuntu_cpu_openblas
 }
 
+build_ubuntu_base_cpu() {
+    echo "skip"
```
Comment (lebeg) [inline ci/dockerfiles/runtime_functions.sh]: Do we need this?

File: ci/build.py (hunk 1 above)
Comment (larroy) [reply ci/build.py]: agree, just trying to see if I could get Jenkins to run. Will remove it before final merge.

File: ci/build.py (hunk 5 above)
Comment (larroy) [reply ci/build.py]: you suggest info?

File: ci/build.py (hunk 11 above)
Comment (larroy) [reply ci/build.py]: To get the output of the container.

File: ci/build.py (hunk 12 above)
Comment (larroy) [reply ci/build.py]: Agree, seconds. TODO.

File: ci/build.py (hunk 13 above)
Comment (larroy) [reply ci/build.py]: I plan to remove this, is just for diagnostics. TODO.

File: ci/dockerfiles/runtime_functions.sh (hunk 14 above)
Comment (larroy) [reply ci/dockerfiles/runtime_functions.sh]: Yes, otherwise build.py -a doesn't work.

File: ci/build.py (hunk 4 above)
Comment (larroy) [reply ci/build.py]: TODO

File: ci/build.py (hunk 3 above)
Comment (larroy) [reply ci/build.py]: TODO

**Comment by marcoabreu** [comment]
Could we temporarily close this PR to avoid any accidental executions on prod CI?

**Comment by larroy** [comment]
@marcoabreu I don't think there would be any problem if it's executed by CI, what are you worried about specifically?

**Comment by larroy** [comment]
@lebeg addressed your comments.

File: ci/build.py (hunk 3 above)
Comment (larroy) [reply ci/build.py]: ⚡

File: ci/build.py (hunk 2 above)
Comment (larroy) [reply ci/build.py]: ⚡

File: ci/build.py (hunk 1 above)
Comment (larroy) [reply ci/build.py]: ⚡

File: ci/build.py (hunk 12 above)
Comment (larroy) [reply ci/build.py]: ⚡

File: ci/build.py (hunk 5 above)
Comment (lebeg) [reply ci/build.py]: Yes, might be more appropriate

File: ci/dockerfiles/runtime_functions.sh (hunk 14 above)
Comment (lebeg) [reply ci/dockerfiles/runtime_functions.sh]: Ok, I see

**Comment by marcoabreu** [comment]
See Antons comment about your change deleting the entire shared CCache, for example. I'm closing now since there are being new pushes to this PR.

File: ci/build.py (hunk 8 above)
Comment (larroy) [reply ci/build.py]: done
